net.rocketpool/image/src/app.ts

```diff
@@ -1,163 +1,240 @@
   blockNumber: number;
   blockHash: string;
-  blockTime: number;
+  eventTime: number;
   transactionIndex: number;
   transactionHash: string;
   logIndex: number;
 }
 
+interface OdfReadRowsResult {
+  etag: number;
+  hasMore: boolean;
+}
+
+type Provider = ethers.providers.JsonRpcProvider | ethers.providers.WebSocketProvider;
+type OnEventCallback = (_: EventRow) => Promise<boolean>;
+type AllBlockRangeHasProcessed = boolean;
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+async function main() {
+  const env = cleanEnv(process.env, {
+    // Application:
+    ETH_NODE_PROVIDER_URL: url(),
+    BLOCK_BATCH_SIZE: num(),
+    // ODF interface:
+    ODF_BATCH_SIZE: num(),
+    ODF_ETAG: num({ default: 0 }),
+    ODF_NEW_ETAG_PATH: str(),
+    ODF_NEW_HAS_MORE_DATA_PATH: str(),
+  })
+
+  // See: https://etherscan.io/token/0xae78736cd615f374d3085123a210448e74fc6393
+  //      https://etherscan.io/address/0xae78736cd615f374d3085123a210448e74fc6393
+  const rethAddress = "0xae78736cd615f374d3085123a210448e74fc6393";
+  const rethStartBlock = 13325304;
+
+  // Init
+  const providerUrl = env.ETH_NODE_PROVIDER_URL;
+
+  console.log(`ETH node provider URL: ${providerUrl}`);
+
+  const provider = getProvider(providerUrl)
+  const rethContract = RethAbi__factory.connect(rethAddress, provider)
+
+  // Block range
+  const lastSeenBlock = env.ODF_ETAG;
+
+  console.log(`Last seen block: ${lastSeenBlock}`);
+
+  const ethHeadBlock = await provider.getBlockNumber();
+
+  console.log(`Current chain head block: ${ethHeadBlock}`);
+
+  const startBlock = Math.max(lastSeenBlock + 1, rethStartBlock);
+  const endBlock = ethHeadBlock;
+  const {
+    etag,
+    hasMore,
+  } = await odfReadRows(rethContract, startBlock, endBlock, env.BLOCK_BATCH_SIZE, env.ODF_BATCH_SIZE);
+
+  // Notify `kamu-cli` regarding new state
+  fs.writeFileSync(env.ODF_NEW_ETAG_PATH, etag.toString());
+
+  if (hasMore) {
+    fs.writeFileSync(env.ODF_NEW_HAS_MORE_DATA_PATH, "");
+  }
+}
+
 /////////////////////////////////////////////////////////////////////////////////////////
 
-async function readEventsFromETH(
-  startBlock,
-  endBlock,
-  clb: (_: Event) => boolean,
-  stride = 10000
-) {
-  console.log(
-    "Considering events in block range [%s, %s] (%s blocks)",
+async function odfReadRows(
+  rethContract: RethAbi,
+  startBlock: number,
+  endBlock: number,
+  blockBatchSize: number,
+  odfBatchSize: number,
+): Promise<OdfReadRowsResult> {
+  let outputRowsCount = 0;
+  let lastOutputBlockNumber = 0;
+
+  const allBlockRangeHasProcessed = await readRethContractEvents(
+    rethContract,
     startBlock,
     endBlock,
-    endBlock - startBlock
+    async function (eventRow) {
+      if (outputRowsCount >= odfBatchSize) {
+        return false;
+      }
+
+      outputRowsCount++;
+      lastOutputBlockNumber = eventRow.blockNumber;
+
+      await stdOutWrite(`${JSON.stringify(eventRow)}\n`);
+
+      return true;
+    },
```

Comment:
According to this logic we can start dropping events mid-block. Upon the next run of the script this may result in emitting duplicate records.

I think we better write an entire batch out before terminating the iteration - i.e. we stop iterating always on block boundaries.

Reply:
Done